components/ai_chat/content/browser/bookmarks_service.cc

```diff
@@ -0,0 +1,151 @@
+  auto id = bookmark_listeners_.Add(std::move(pending_listener));
+  auto* listener = bookmark_listeners_.Get(id);
+  if (listener) {
+    listener->Changed(GetAllBookmarks());
+  }
+}
+
+void BookmarksService::BookmarkModelLoaded(bool ids_reassigned) {
+  auto change = GetAllBookmarks();
+  for (const auto& listener : bookmark_listeners_) {
+    listener->Changed(change->Clone());
+  }
+}
+
+void BookmarksService::BookmarkNodeAdded(const bookmarks::BookmarkNode* parent,
+                                         size_t index,
+                                         bool added_by_user) {
+  if (!parent || index >= parent->children().size()) {
+    return;
+  }
+
+  auto* node = parent->children()[index].get();
+  if (!node || node->is_folder()) {
+    return;
+  }
+
+  auto change = mojom::BookmarksChange::New();
+  change->addedOrUpdated[node->url().spec()] = ToMojoBookmark(node);
+
+  for (const auto& listener : bookmark_listeners_) {
+    listener->Changed(change->Clone());
+  }
+}
+
+void BookmarksService::BookmarkNodeRemoved(
+    const bookmarks::BookmarkNode* parent,
+    size_t old_index,
+    const bookmarks::BookmarkNode* node,
+    const std::set<GURL>& no_longer_bookmarked,
+    const base::Location& location) {
+  if (node->is_folder()) {
+    return;
+  }
+
+  auto change = mojom::BookmarksChange::New();
+  change->removed.push_back(node->url().spec());
+
+  for (const auto& listener : bookmark_listeners_) {
+    listener->Changed(change->Clone());
+  }
+}
+
+void BookmarksService::BookmarkNodeChanged(
+    const bookmarks::BookmarkNode* node) {
+  if (node->is_folder()) {
+    return;
+  }
+  auto change = mojom::BookmarksChange::New();
+  change->addedOrUpdated[node->url().spec()] = ToMojoBookmark(node);
+
+  for (const auto& listener : bookmark_listeners_) {
+    listener->Changed(std::move(change));
+  }
+}
+
+mojom::BookmarksChangePtr BookmarksService::GetAllBookmarks() {
+  auto change = mojom::BookmarksChange::New();
+  if (!bookmark_model_) {
+    return change;
+  }
+
+  // Do a depth-first traversal of the bookmark tree extracting all non-folder
+  // nodes.
+  std::vector<const bookmarks::BookmarkNode*> frontier;
+  frontier.push_back(bookmark_model_->root_node());
+
+  while (!frontier.empty()) {
+    auto* node = frontier.back();
+    frontier.pop_back();
+
+    if (node->is_folder()) {
+      auto& children = node->children();
+      std::ranges::transform(children.begin(), children.end(),
+                             std::back_inserter(frontier),
+                             [](const auto& c) { return c.get(); });
+    } else {
+      change->addedOrUpdated[node->url().spec()] = ToMojoBookmark(node);
+    }
+  }
+
+  return change;
+}
+
+void BookmarksService::BookmarkAllUserNodesRemoved(
+    const std::set<GURL>& removed_urls,
+    const base::Location& location) {
+  // Refresh all bookmarks since all user nodes were removed
+  auto change = GetAllBookmarks();
```

Comment:
why isn't it `auto change = mojom::BookmarksChange::New();` since all bookmarks are gone?

Reply:
addressed in https://github.com/brave/brave-core/pull/30963/commits/6a20bdc50e39ab2ca5cafdb9eaedee3b328c0895